bluetooth: host: a2dp: error_code should be detailed in set_config_rsp

CC0918 opened the pull request:
Error codes fall into two main categories:
case 1: NOT_SUPPORT_XXX. This means that the configuration of the parameter is not supported locally.
case 2: INVALID_XXX. This means that the parameter is not configured or duplicated in the set_config_req command.

Currently the protocol only supports SBC, so it is only tested for SBC parameters.

MarkWangChinese commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -148,6 +148,20 @@ static int a2dp_get_capabilities_ind(struct bt_avdtp *session, struct bt_avdtp_s
 	return 0;
 }
 
+#define DUPLICATE_BIT(config) (((config) & ((config )- 1)) != 0)
```
suggest to change as: IS_BIT_DUPLICATED(val)

CC0918 replied:
update

CC0918 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -190,14 +204,55 @@ static int a2dp_process_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep
+		return -EINVAL;
+	} else {
+		*errcode = BT_A2DP_INVALID_CODEC_TYPE;
+		return -EINVAL;
```
Done

CC0918 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -190,14 +204,55 @@ static int a2dp_process_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep
+			*errcode = BT_A2DP_INVALID_MAXIMUM_BITPOOL_VALUE;
+			return -EINVAL;
+		} else if (sbc_set->max_bitpool > sbc->max_bitpool) {
+			*errcode = BT_A2DP_NOT_SUPPORTED_MAXIMUM_BITPOOL_VALUE;
```
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -179,6 +259,11 @@ static int a2dp_process_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep
 	}
 
+	*errcode = a2dp_check_codec_type(ep, codec_type);
+	if (*errcode) {
```
```suggestion
	if (*errcode != 0) {
```

CC0918 replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -190,16 +275,12 @@ static int a2dp_process_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep
-			*errcode = BT_AVDTP_BAD_ACP_SEID;
+
+		*errcode = a2dp_check_config_param(sbc_set, sbc);
+		if (*errcode) {
```
```suggestion
	if (*errcode != 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -148,6 +148,86 @@ static int a2dp_get_capabilities_ind(struct bt_avdtp *session, struct bt_avdtp_s
+{
+	uint8_t errcode = 0;
+
+	if (codec_type != BT_A2DP_SBC && codec_type != BT_A2DP_MPEG1 &&
+	    codec_type != BT_A2DP_MPEG2 && codec_type != BT_A2DP_ATRAC &&
+	    codec_type != BT_A2DP_VENDOR) {
```
The codec ID `MPEG-D USAC` is missing.

CC0918 commented on subsys/bluetooth/host/classic/a2dp.c:
```diff
@@ -148,6 +148,88 @@ static int a2dp_get_capabilities_ind(struct bt_avdtp *session, struct bt_avdtp_s
+	} else {
+	}
+
+	return errcode;
```
Done

CC0918 committed:
bluetooth: host: a2dp:error_code should be detailed in set_config_rsp

Error codes fall into two main categories:
case 1: NOT_SUPPORT_XXX. This means that the configuration of the
parameter is not supported locally.
case 2: INVALID_XXX. This means that the parameter is not configured or
duplicated in the set_config_req command.

Currently the protocol only supports SBC, so it is only checked for SBC
parameters.

Signed-off-by: Cheng Chang <cheng.chang@nxp.com>
